# Deep mutational scanning of the Nipah virus fusion protein (F) using a barcoded lentiviral vector
[![Documentation](https://img.shields.io/badge/docs-GitHub%20Pages-blue)](https://dms-vep.github.io/Nipah_Malaysia_F_DMS/)
[![DOI](https://img.shields.io/badge/DOI-Preprint-purple)](https://www.biorxiv.org/content/10.1101/2025.10.15.682664v1)

A deep mutational scanning study of the Nipah virus fusion protein to characterize mutations affecting cell entry and antibody neutralization using lentiviral pseudotyping.


## Table of Contents
- [Key Data Files](#key-data-files)
- [Project Information](#project-information)
- [Repository Structure](#repository-structure)
- [Usage](#usage)
- [Biosafety](#biosafety)
- [Supplemental Files](#supplemental-files)

## Key Data Files
The main outputs from this study are available as filtered datasets:

- **[Cell entry data](analysis/results/filtered_data/cell_entry/Nipah_F_func_effects_filtered.csv)** - Functional effects of mutations on viral entry
- **[Antibody escape data](analysis/results/filtered_data/antibody_escape/combined/all_antibodies_escape_filtered.csv)** - Mutation effects on antibody neutralization


## Project Information
This repo contains data and analyses from deep mutational scanning experiments on the Nipah virus fusion (F) protein. All experiments were performed on a [codon-optimized](data/paper_reference_files/sequences/unmutated_reference/NipahF_codon_optimized_CTdel_DNA.fasta) F sequence. Twenty-two amino acids were removed from the cytoplasmic tail to improve lentivirus titers, and all residues between sites 29 and 481 contain every possible single missense mutation. Stop codons were designed to occur at every other position from positions 29 to 104.


## Repository Structure
The analysis is mostly run by [dms-vep-pipeline-3](https://github.com/dms-vep/dms-vep-pipeline-3), which is included as a [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules).

The configuration for the analysis is in [config.yaml](config.yaml), and additional input data are in [./data/](data).

The results of running the pipeline are placed in [./results/](results), only some of which are tracked per the [.gitignore](.gitignore) file.
HTML documentation is created by the pipeline and placed in [./docs/](docs), where it is rendered via GitHub Pages at [https://dms-vep.github.io/Nipah_Malaysia_F_DMS/](https://dms-vep.github.io/Nipah_Malaysia_F_DMS/)

There is also an additional snakemake pipeline in [analysis/](analysis) that runs python scripts to filter the data generated by the [dms-vep-pipeline-3](https://github.com/dms-vep/dms-vep-pipeline-3) pipeline, and to generate the final data files and figures used in the manuscript. The output of these scripts are placed in [analysis/results/](analysis/results).

## Usage
To clone the repo, run:
    
    git clone https://github.com/dms-vep/Nipah_Malaysia_F_DMS.git

To run the analysis, create the `conda` environment in [dms-vep-pipeline-3/environment.yml](dms-vep-pipeline-3/environment.yml), activate it, and run the `snakemake` pipeline in [dms-vep-pipeline-3/Snakefile](dms-vep-pipeline-3/Snakefile) with:
    
    conda activate dms-vep-pipeline-3

    snakemake -s dms-vep-pipeline-3/Snakefile -j <n_jobs>
    
On the Hutch cluster, you can also run with the `slurm` script [run_Hutch_cluster.bash](run_Hutch_cluster.bash) with:

    sbatch run_Hutch_cluster.bash

## Biosafety
All experiments were performed with non-replicative lentiviral-based pseudoviruses in a biosafety-level 2 laboratory by trained personnel. Pseudotyping is a method where viral entry proteins are expressed in combination with a viral vector from a different virus. By only expressing the Nipah receptor binding and fusion proteins on the surface of lentiviral particles, we can safely perform deep mutational scanning experiments without modifying authentic virus. Essential lentiviral genes, such as gag/pol, rev, and tat, are not encoded by the lentiviral vector. These genes are instead introduced by transfecting cells with three separate plasmids. This ensures the pseudotyped lentiviruses cannot replicate outside of a cell culture system where these plasmids are co-transfected.


## Supplemental Files
Supplemental data for the manuscript are located in the [data/paper_reference_files/sequences](data/paper_reference_files/sequences) directory. This includes:
- `designed_oPools/`: Contains the DNA sequences of the designed oligonucleotide pools ordered from Twist Biosciences.
- `destination_vectors/`: Contains plasmid maps of the six destination vectors used for cloning the mutagenized tiles into the Nipah F gene.
- `genbank/`: Contains GenBank accession numbers for sequences used in the manuscript.
- `plasmids/`: Contains maps of plasmids used in the study.
- `primers/`: Contains sequences of primers used in the study
- `unmutated_reference/`: Contains the DNA sequence of the codon-optimized, CT-truncated Nipah F gene used as the parental sequence for mutagenesis.

