{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9e0143c4-e8a2-4a0e-bfcb-f724b8947ceb",
   "metadata": {},
   "source": [
    "# Make oligo sequences for DMS library\n",
    "Script written by Brendan Larsen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3bc6507-4fb3-4dea-b078-ec641ba85402",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import AlignIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e165321-fa5a-485e-9c6b-27e586f7258b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define all codons\n",
    "full_codon_table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af06bdb-ac41-4af0-8d45-fc061b8f45e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## https://www.genscript.com/tools/codon-frequency-table\n",
    "## Define the codons we want to use for each mutant, depending on human codon frequencies.\n",
    "## These are the most frequent codons for each amino acid in humans:\n",
    "\n",
    "codon_table = {\n",
    "    'GCC':'A',\n",
    "    'TGC':'C',\n",
    "    'GAT':'D', #change from GAC because can make BsmBI sites\n",
    "    'GAG':'E',\n",
    "    'TTC':'F',\n",
    "    'GGC':'G',\n",
    "    'CAC':'H',\n",
    "    'ATC':'I',\n",
    "    'AAG':'K',\n",
    "    'CTG':'L',\n",
    "    'ATG':'M',\n",
    "    'AAC':'N',\n",
    "    'CCC':'P',\n",
    "    'CAG':'Q',\n",
    "    'CGG':'R',\n",
    "    'AGC':'S',\n",
    "    'ACC':'T',\n",
    "    'GTG':'V',\n",
    "    'TGG':'W',\n",
    "    'TAC':'Y',\n",
    "    'TAA':'*' \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54681361-5b40-45f1-a715-78e7a68fcf93",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_mutations(\n",
    "    input_file,\n",
    "    output_file,\n",
    "    fwd_primer='',\n",
    "    fwd_spacer='',\n",
    "    fwd_overhang='',\n",
    "    fwd_additional='',\n",
    "    rev_additional='',\n",
    "    rev_overhang='',\n",
    "    rev_spacer='',\n",
    "    rev_primer='',\n",
    "    primer_buffer='',\n",
    "    include_stop_codons=False,\n",
    "    stop_codons_every_other=False\n",
    "):\n",
    "    \"\"\"\n",
    "    Generate mutations for a given DNA sequence.\n",
    "\n",
    "    Parameters:\n",
    "    - input_file (str): Path to the input FASTA file containing the sequence to mutate.\n",
    "    - output_file (str): Path to save the output FASTA file with mutations.\n",
    "    - fwd_primer (str, optional): Forward primer sequence.\n",
    "    - fwd_spacer (str, optional): Forward spacer sequence.\n",
    "    - fwd_overhang (str, optional): Forward overhang sequence.\n",
    "    - fwd_additional (str, optional): Additional forward sequence.\n",
    "    - rev_additional (str, optional): Additional reverse sequence.\n",
    "    - rev_overhang (str, optional): Reverse overhang sequence.\n",
    "    - rev_spacer (str, optional): Reverse spacer sequence.\n",
    "    - rev_primer (str, optional): Reverse primer sequence.\n",
    "    - primer_buffer (str, optional): Primer buffer sequence.\n",
    "    - include_stop_codons (bool, optional): Whether to include stop codons in mutations. Default is True.\n",
    "    - stop_codons_every_other (bool, optional): If True, include stop codons at every other site. Default is False.\n",
    "    \"\"\"\n",
    "\n",
    "    # Read Input Amplicon\n",
    "    seq = Seq(str(AlignIO.read(input_file, 'fasta')[0].seq))\n",
    "    \n",
    "    # Golden Gate recognition sites\n",
    "    bsmbi_fwd = 'CGTCTC'\n",
    "    bsmbi_rev = 'GAGACG'\n",
    "    \n",
    "    # Make empty list to append mutations to\n",
    "    mutations = []\n",
    "    \n",
    "    # Make the wildtype sequence at the top of the file\n",
    "    wt_seq = primer_buffer + fwd_primer + bsmbi_fwd + fwd_spacer + fwd_overhang + fwd_additional + seq + rev_additional + rev_overhang + rev_spacer + bsmbi_rev + rev_primer + primer_buffer\n",
    "    mutations.append(SeqRecord(wt_seq, id='WT_seq'))\n",
    "    \n",
    "    # Loop through each position and make all possible mutations based on codon table\n",
    "    num_codons = len(seq) // 3\n",
    "    for codon_index in range(num_codons):\n",
    "        codon_start = codon_index * 3\n",
    "        codon_end = codon_start + 3\n",
    "        original_codon = seq[codon_start:codon_end]\n",
    "        original_aa = full_codon_table[str(original_codon)]\n",
    "        \n",
    "        for mutated_codon, amino_acid in codon_table.items():\n",
    "            # Skip if the mutated amino acid is the same as the original amino acid\n",
    "            if amino_acid != original_aa:\n",
    "                # Check stop codon conditions\n",
    "                if (amino_acid != \"*\" or  # Always include non-stop codon mutations\n",
    "                    (include_stop_codons and  # Include stop codons if specified\n",
    "                     (not stop_codons_every_other or  # Either include all stop codons\n",
    "                      (stop_codons_every_other and codon_index % 2 == 0)))):  # Or include stop codons at every other position\n",
    "                    mutated_seq = (primer_buffer + fwd_primer + bsmbi_fwd + fwd_spacer + \n",
    "                                   fwd_overhang + fwd_additional + \n",
    "                                   seq[:codon_start] + Seq(mutated_codon) + seq[codon_end:] + \n",
    "                                   rev_additional + rev_overhang + rev_spacer + \n",
    "                                   bsmbi_rev + rev_primer + primer_buffer)\n",
    "                    mutations.append(SeqRecord(mutated_seq, id=f\"Pool1_Mutation_{codon_index+1}_{amino_acid}\", description=\"\"))\n",
    "    \n",
    "    # Save the mutations as a FASTA file\n",
    "    SeqIO.write(mutations, output_file, \"fasta\")\n",
    "    print(f\"Mutations saved to {output_file}\")\n",
    "    print(f\"Total mutations generated: {len(mutations) - 1}\")  # Subtract 1 to exclude the WT sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86ca51d4-d960-4281-b00a-f8ebc4873b3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#amplicon1\n",
    "generate_mutations(\n",
    "    input_file='amplicons/pool1.fasta',\n",
    "    output_file='output/amplicon1_oPools.fasta',\n",
    "    \n",
    "    primer_buffer='TA',\n",
    "    fwd_primer='CCAGGCAGCATGTTGTAGTG',\n",
    "    fwd_spacer='A',\n",
    "    fwd_overhang='TCCT',\n",
    "    fwd_additional='G',\n",
    "\n",
    "    rev_additional='GTG',\n",
    "    rev_overhang='GGCG',\n",
    "    rev_spacer='A',\n",
    "    rev_primer='GTACGAGCCGATGGAATCAG',\n",
    "\n",
    "    include_stop_codons=True,\n",
    "    stop_codons_every_other=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c93fbe80-3f61-4d3b-913e-1da8e748cd13",
   "metadata": {},
   "outputs": [],
   "source": [
    "#amplicon2\n",
    "generate_mutations(\n",
    "    input_file='amplicons/pool2.fasta',\n",
    "    output_file='output/amplicon2_oPools.fasta',\n",
    "    \n",
    "    primer_buffer='ATA',\n",
    "    fwd_primer='TGGAGAGACTACGCACTGAC',\n",
    "    fwd_spacer='A',\n",
    "    fwd_overhang='CACG',\n",
    "    fwd_additional='AC',\n",
    "\n",
    "    rev_additional='AT',\n",
    "    rev_overhang='CAAC',\n",
    "    rev_spacer='T',\n",
    "    rev_primer='GCACAGTGTACGAGATGGTC',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eba2f854-5638-4f08-a889-ae1f40ced255",
   "metadata": {},
   "outputs": [],
   "source": [
    "#amplicon3\n",
    "generate_mutations(\n",
    "    input_file='amplicons/pool3.fasta',\n",
    "    output_file='output/amplicon3_oPools.fasta',\n",
    "    \n",
    "    primer_buffer='AA',\n",
    "    fwd_primer='TCAAGCACTGCACTCAAGTC',\n",
    "    fwd_spacer='A',\n",
    "    fwd_overhang='GGAC',\n",
    "    fwd_additional='',\n",
    "\n",
    "    rev_additional='G',\n",
    "    rev_overhang='ATCT',\n",
    "    rev_spacer='C',\n",
    "    rev_primer='CCTCACTCGTCACTACCATG',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b35470fe-3cf8-4d26-8efd-0235c796b50d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#amplicon4\n",
    "generate_mutations(\n",
    "    input_file='amplicons/pool4.fasta',\n",
    "    output_file='output/amplicon4_oPools.fasta',\n",
    "    \n",
    "    primer_buffer='TTA',\n",
    "    fwd_primer='TCATCCATCCAGTACGCCAG',\n",
    "    fwd_spacer='A',\n",
    "    fwd_overhang='ACCG',\n",
    "    fwd_additional='AG',\n",
    "\n",
    "    rev_additional='A',\n",
    "    rev_overhang='TCGG',\n",
    "    rev_spacer='A',\n",
    "    rev_primer='CAGCAGCATACGTCGGTAAC',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "916a3a8b-9bcb-4087-8c73-cf6de390faa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#amplicon5\n",
    "generate_mutations(\n",
    "    input_file='amplicons/pool5.fasta',\n",
    "    output_file='output/amplicon5_oPools.fasta',\n",
    "    \n",
    "    primer_buffer='',\n",
    "    fwd_primer='CGAGTCCAGTTACGGAGATG',\n",
    "    fwd_spacer='C',\n",
    "    fwd_overhang='TATC',\n",
    "    fwd_additional='',\n",
    "\n",
    "    rev_additional='GG',\n",
    "    rev_overhang='CGAA',\n",
    "    rev_spacer='T',\n",
    "    rev_primer='CTGTCTTGCGTCCATCCATG',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "639b39c1-8334-451a-a60a-ec5cdfce482d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#amplicon6\n",
    "generate_mutations(\n",
    "    input_file='amplicons/pool6.fasta',\n",
    "    output_file='output/amplicon6_oPools.fasta',\n",
    "\n",
    "    primer_buffer='',\n",
    "    fwd_primer='AGGTTAGACAGGCTGACGTG',\n",
    "    fwd_spacer='A',\n",
    "    fwd_overhang='CTCA',\n",
    "    fwd_additional='G',\n",
    "\n",
    "    rev_additional='G',\n",
    "    rev_overhang='ACAC',\n",
    "    rev_spacer='T',\n",
    "    rev_primer='GCTGTCATCACCTCAACTGC',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa7c56b8-50a8-4bc2-a2e4-8beaf9685730",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
